Nodejs/Product/TestAdapter/TestExecutor.cs

```diff
@@ -235,16 +269,52 @@ select connection.LocalEndPoint.Port
                     }
 #endif
                 }
+                // Send the process the list of tests to run and wait for it to complete
+                _nodeProcess.WriteInputLine(JsonConvert.SerializeObject(testObjects));
+                _nodeProcess.Wait();
+
+                // Automatically fail tests that haven't been run by this point (failures in before() hooks)
+                foreach(TestCase notRunTest in _currentTests) {
+                    TestResult result = new TestResult(notRunTest);
+                    result.Outcome = TestOutcome.Failed;
+                    if(_currentResultObject != null) {
+                        result.Messages.Add(new TestResultMessage(TestResultMessage.StandardOutCategory, _currentResultObject.stdout));
+                        result.Messages.Add(new TestResultMessage(TestResultMessage.StandardErrorCategory, _currentResultObject.stderr));
+                    }
+                    frameworkHandle.RecordResult(result);
+                    frameworkHandle.RecordEnd(notRunTest, TestOutcome.Failed);
+                }
             }
+        }
 
-            WaitHandle.WaitAll(new WaitHandle[] { _nodeProcess.WaitHandle });
+        private void KillNodeProcess() {
+            lock (_syncObject) {
+                if (_nodeProcess != null) {
+                    _nodeProcess.Kill();
+                }
+            }
+        }
 
-            bool runCancelled = _cancelRequested.WaitOne(0);
-            RecordEnd(frameworkHandle, test, testResult,
-                string.Join(Environment.NewLine, _nodeProcess.StandardOutputLines),
-                string.Join(Environment.NewLine, _nodeProcess.StandardErrorLines),
-                (!runCancelled && _nodeProcess.ExitCode == 0) ? TestOutcome.Passed : TestOutcome.Failed);
-            _nodeProcess.Dispose();
+        private static int GetFreePort() {
```

Comment:
Can you add a note where you copied this from ( I've seen something similar in the debugger code).. I think we want to refactor in the future to clean that up.